src/utils/io.c

```diff
@@ -0,0 +1,55 @@
+
+#include <string.h>
+
+#include "utils/io.h"
+#include "utils/log.h"
+#include "utils/miscmath.h"
+
+#define READ_BLOCK_SIZE 32
+
+FILE *file_open(const char *file_name, const char *mode) {
+    FILE *handle = fopen(file_name, mode);
+    if(handle == NULL) {
+        PERROR("Unable to open file: %s: %d", file_name, strerror(errno));
+        abort();
+    }
+    return handle;
+}
+
+void file_seek(FILE *handle, long offset, int origin) {
+    if(fseek(handle, offset, origin)) {
+        PERROR("Unable to fseek to %d %d: %s", origin, offset, strerror(errno));
+        abort();
+    }
+}
+
+long file_size(FILE *handle) {
+    long pos, file_size;
+    pos = ftell(handle);
+    file_seek(handle, 0, SEEK_END);
+    file_size = ftell(handle);
+    file_seek(handle, pos, SEEK_SET);
+    return file_size;
+}
+
+void file_read(FILE *handle, char *buffer, long size) {
+    size_t ptr = 0;
+    size_t read_size;
+    while(1) {
```

Comment:
Why this complicated crap? Just `fread` once with the correct size and error out if it fails.

Reply:
Because if I read with anything over 1024, fread fails to EINVAL (invalid argument).

Reply:
Also, technically it is possible for fread to return less then <count> amount of elements, while returning more on subsequent reads. It's valid behaviour for fread.

Reply:
I have never seen that. What kind of a fucked up platform are you using?

Reply:
Windows 10 + MSYS2 (mingw) triggered this one. I don't think you can get much more fucked up than that ?

Reply:
And yes, works fine on ubuntu before and now.

Reply:
At least 1. put some comments on it and 2. use an optimistic strategy so non-broken platforms aren't slowed down.

Reply:
Not going to touch this anymore otherwise. I added the rest of the work on my todo list for cleanup.